File: src/context/token-counter.ts
import { encodingForModel } from "js-tiktoken";

export interface ContextMessage {
  role: "system" | "user" | "assistant";
  content: string;
}

const MESSAGE_OVERHEAD_TOKENS = 6;

let encoder: ReturnType<typeof encodingForModel> | null = null;

function getEncoder(): ReturnType<typeof encodingForModel> {
  if (!encoder) {
    encoder = encodingForModel("gpt-4");
  }

  return encoder;
}

export function estimateTextTokens(text: string): number {
  if (text.length === 0) {
    return 0;
  }

  return getEncoder().encode(text).length;
}

export function estimateMessageTokens(message: ContextMessage): number {
  return MESSAGE_OVERHEAD_TOKENS + estimateTextTokens(message.content);
}

export function estimateMessagesTokens(messages: ContextMessage[]): number {
  return messages.reduce((total, message) => total + estimateMessageTokens(message), 0);
}

export function estimateLooseMessageTokens(message: { content?: string | null }): number {
  return MESSAGE_OVERHEAD_TOKENS + estimateTextTokens(message.content ?? "");
}

File: internal/context/token_counter.go
package context

import "unicode/utf8"

const messageOverheadTokens = 6

type Message struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

func EstimateTextTokens(text string) int {
	if text == "" {
		return 0
	}

	runes := utf8.RuneCountInString(text)
	if runes < 4 {
		return 1
	}
	return (runes + 3) / 4
}

func EstimateMessageTokens(message Message) int {
	return messageOverheadTokens + EstimateTextTokens(message.Content)
}

func EstimateMessagesTokens(messages []Message) int {
	total := 0
	for _, message := range messages {
		total += EstimateMessageTokens(message)
	}
	return total
}
